agent/hcp/manager.go

```diff
@@ -71,15 +77,80 @@ func NewManager(cfg ManagerConfig) *Manager {
 	}
 }
 
+// runReporter initializes, and if successful, then runs the metrics reporter.
+func (m *Manager) runReporter(ctx context.Context) error {
+	// Step 1: Obtain CCM telemetry configuration
+	// Only enable HCP metrics reporting if server is registered with management plane.
+	telemetryCfg, err := m.cfg.Client.FetchTelemetryConfig(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to obtain CCM telemetry config: %v", err)
+	}
+
+	if telemetryCfg == nil {
+		// Not an error, as the server is not registered with CCM, but early return.
+		return nil
+	}
+
+	// Step 2: Init telemetry.MetricsExporter which sends metrics to HCP Metrics Gateway in OTLP format.
+	// It uses a an OTLP exporter client wrapped by the HCP client.
+	// This enables us to perform HCP auth and easily mock the client interface for tests.
+	// It must first be initialized within the HCP Client with the configured endpoint.
+	if err := m.cfg.Client.InitMetricsClient(ctx, telemetryCfg.Endpoint); err != nil {
+		return fmt.Errorf("failed to init metrics HCP client: %v", err)
+	}
+
+	expCfg := &telemetry.MetricsExporterConfig{
+		Labels: map[string]string{
+			"service.name":        "consul-server",
+			"service.version":     version.GetHumanVersion(),
+			"service.instance.id": m.cfg.NodeID,
+		},
+		Logger:  m.logger.Named("metrics_exporter"),
+		Filters: telemetryCfg.Filters,
+		// Inject client in metrics exporter.
+		Client: m.cfg.Client,
+	}
+
+	exp, err := telemetry.NewMetricsExporter(expCfg)
+	if err != nil {
+		return fmt.Errorf("failed to create exporter: %v", err)
+	}
+
+	// Step 3: Init telemetry.Reporter, which gathers consul server go metrics over a configurable time interval (Report Interval).
+	// It flushes them to the exporter to be sent to HCP at a configurable time interval (Batch Interval).
+	cfg := telemetry.DefaultConfig()
+	cfg.Logger = m.logger.Named("telemetry_reporter")
+	cfg.Gatherer = m.cfg.MetricsBackend
+	cfg.Exporter = exp
+
+	m.reporter, err = telemetry.NewReporter(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create exporter: %v", err)
+	}
+
+	// If setup is successful, run the reporter, which is a blocking operation.
+	m.reporter.Run(ctx)
+
+	return nil
+}
+
 // Run executes the Manager it's designed to be run in its own goroutine for
 // the life of a server agent. It should be run even if HCP is not configured
 // yet for servers since a config update might configure it later and
 // UpdateConfig called. It will effectively do nothing if there are no HCP
 // credentials set other than wait for some to be added.
 func (m *Manager) Run(ctx context.Context) {
-	var err error
 	m.logger.Debug("HCP manager starting")
 
```

Comment:
One thing that worries me is that we are not handling goroutine exits, in the case that the if the shared context is cancelled both the manager `Run` and this inner goroutine will return with a parent that exited. 

Some thoughts:
- Maybe this is this fine as is? 
- If not:
  - Should I use a separate context for the reporter goroutine, and cancel it when the manager `Run` function cancels?
  - Should I convert the contents of the manager `Run` function into a goroutine `run` and wait on both `runReporter` and `run` ?